### PR TITLE
feat: add ability to create atlas search indexes MCP-275

### PR DESCRIPTION
**Pull request by nirinchev**
## Proposed changes

This adds support for creating atlas search indexes. Dropping and listing is already supported, so needed no changes.

Notable omissions from the schema, that we will want to evaluate based on user feedback and internal testing:
* `analyzers` - based on consultation with the Search team
* `mappings.dynamic.typeSet` - preview feature
* `searchAnalyzer` - based on consultation with the Search team
* `storedSource` - based on testing of model performance
* `synonyms` - for simplicity
* `typeSets` - for simplicity
* `mappings.fields.*` - the fields definition only includes the type and uses `passthrough` for the remaining fields. We'll have to evaluate its performance and if necessary, define the complete json schema.

### CODE REVIEW

File: src/tools/mongodb/create/createIndex.ts [hunk 1]
```diff
@@ -6,61 +6,121 @@ import type { IndexDirection } from "mongodb";
 import { quantizationEnum, similarityEnum } from "../../../common/search/vectorSearchEmbeddingsManager.js";
 
 export class CreateIndexTool extends MongoDBToolBase {
-    private vectorSearchIndexDefinition = z.object({
```
Comment (nirinchev) [inline src/tools/mongodb/create/createIndex.ts:10]: This is best viewed with the "Hide whitespace" option - it's just prettier reformatting the indents.

File: tests/accuracy/sdk/accuracyTestingClient.ts [hunk 2]
```diff
@@ -91,7 +91,7 @@ export class AccuracyTestingClient {
             return [`--${key}`, value];
         });
 
-        const args = [MCP_SERVER_CLI_SCRIPT, "--connectionString", mdbConnectionString, ...additionalArgs];
+        const args = [MCP_SERVER_CLI_SCRIPT, mdbConnectionString, ...additionalArgs];
```
Comment (nirinchev) [inline tests/accuracy/sdk/accuracyTestingClient.ts:95]: `--connectionString` is deprecated so using the positional argument.

File: tests/integration/tools/mongodb/delete/dropIndex.test.ts [hunk 3]
```diff
@@ -362,35 +376,39 @@ describe.each([{ vectorSearchEnabled: false }, { vectorSearchEnabled: true }])(
...
+                                    namespace: `${database}.${collection}`,
                                 });
-                            }
-                        );
+                            });
+                        });
 
-                        describe("and attempting to delete an existing index", () => {
+                        describe.each([
+                            { description: "search", indexName: getSearchIndexName() },
```
Comment (nirinchev) [inline tests/integration/tools/mongodb/delete/dropIndex.test.ts]: This should pass in the function rather than evaluating it

File: src/tools/mongodb/create/createIndex.ts [hunk 4]
```diff
@@ -6,61 +6,121 @@ import type { IndexDirection } from "mongodb";
...
+                .refine((fields) => fields.some((f) => f.type === "vector"), {
+                    message: "At least one vector field must be defined",
+                })
+                .describe(
+                    "Definitions for the vector and filter fields to index, one definition per document. You must specify `vector` for fields that contain vector embeddings and `filter` for additional fields to filter on. At least one vector-type field definition is required."
+                ),
+        })
+        .describe("Definition for a Vector Search index.");
+
+    private atlasSearchIndexDefinition = z
```
Comment (kmruiz) [inline src/tools/mongodb/create/createIndex.ts:69]: Why aren't we supporting custom analyzers?

https://www.mongodb.com/docs/atlas/atlas-search/analyzers/custom/

File: src/tools/mongodb/create/createIndex.ts [hunk 5]
```diff
@@ -6,61 +6,121 @@ import type { IndexDirection } from "mongodb";
...
+        .object({
+            type: z.literal("search"),
+            analyzer: z
```
Comment (kmruiz) [inline src/tools/mongodb/create/createIndex.ts:72]: Probably this should be an enum of the analyzers.

File: src/tools/mongodb/create/createIndex.ts [hunk 6]
```diff
@@ -6,61 +6,121 @@ import type { IndexDirection } from "mongodb";
...
+                .string()
+                .optional()
+                .default("lucene.standard")
+                .describe(
+                    "The analyzer to use for the index. Can be one of the built-in lucene analyzers (`lucene.standard`, `lucene.simple`, `lucene.whitespace`, `lucene.keyword`), a language-specific analyzer, such as `lucene.cjk` or `lucene.czech`, or a custom analyzer defined in the Atlas UI."
+                ),
+            mappings: z
+                .object({
```
Comment (kmruiz) [inline src/tools/mongodb/create/createIndex.ts:80]: Lack support of:

* numPartitions
* searchAnalyzer vs analyze
* custom analyzers
* storedSources
* synonyms
* typeSets

Comment (kmruiz) [reply src/tools/mongodb/create/createIndex.ts:80]: https://www.mongodb.com/docs/atlas/atlas-search/index-definitions/?deployment-type=atlas&interface=driver&language=nodejs#std-label-ref-index-definitions

We could say that custom analyzers are not that important, but storedSources is actually relevant most of the times.

File: src/tools/mongodb/create/createIndex.ts [hunk 7]
```diff
@@ -6,61 +6,121 @@ import type { IndexDirection } from "mongodb";
...
+                    dynamic: z
+                        .boolean()
```
Comment (kmruiz) [inline src/tools/mongodb/create/createIndex.ts:82]: Dynamic can be an object of typeSets.

File: src/tools/mongodb/create/createIndex.ts [hunk 8]
```diff
@@ -6,61 +6,121 @@ import type { IndexDirection } from "mongodb";
...
+                        .default(false)
+                        .describe(
+                            "Enables or disables dynamic mapping of fields for this index. If set to true, Atlas Search recursively indexes all dynamically indexable fields. If set to false, you must specify individual fields to index using mappings.fields."
+                        ),
+                    fields: z
+                        .record(
+                            z.string().describe("The field name"),
+                            z
+                                .object({
+                                    type: z
```
Comment (kmruiz) [inline src/tools/mongodb/create/createIndex.ts:93]: Objects will require additional fields depending on the type. I know passthrough will keep them, but we should document them so the agent knows which ones to use and how. For example, autocomplete supports defining a custom analyzer, how to tokenize (which is really important) and similarity functions.

File: src/tools/mongodb/create/createIndex.ts (hunk 6 above)
Comment (nirinchev) [reply src/tools/mongodb/create/createIndex.ts:80]: The args shape is based on the POC the search team did for index support and I was going off of the assumption that they've selected the fields that they see the most value in exposing to LLMs. I realize there's a lot more configuration that's possible, I'm just not sure how much of that is stuff we expect agents to configure vs an actual human who wants to fine-tune the index.

File: src/tools/mongodb/create/createIndex.ts (hunk 7 above)
Comment (nirinchev) [reply src/tools/mongodb/create/createIndex.ts:82]: This is in preview though, so I don't expect there's sufficient docs or training data for general-purpose models to accurately choose which one to use.

File: src/tools/mongodb/create/createIndex.ts (hunk 8 above)
Comment (nirinchev) [reply src/tools/mongodb/create/createIndex.ts:93]: The exact shape is extremely complex to represent in a json schema. I'm worried that being overly specific will result in this being more harmful than helpful, especially if we expect the majority of the use cases to revolve around just specifying the type.

File: src/tools/mongodb/create/createIndex.ts (hunk 6 above)
Comment (kmruiz) [reply src/tools/mongodb/create/createIndex.ts:80]: A POC to see the feasibility to create search indexes and production code are likely to have different requirements.

File: src/tools/mongodb/create/createIndex.ts (hunk 7 above)
Comment (kmruiz) [reply src/tools/mongodb/create/createIndex.ts:82]: The preview is for vector search though, FTS is explicitly out of scope of the vector search project.

File: src/tools/mongodb/create/createIndex.ts (hunk 8 above)
Comment (kmruiz) [reply src/tools/mongodb/create/createIndex.ts:93]: Yep, the schema is complicated, it has a lot of options that are not compatible even between them. We should have proper documentation of which ones we want to expose and which ones not, something that we haven't discussed yet because supporting the most used bits of Atlas Search is already a substantial effort.

File: src/tools/mongodb/create/createIndex.ts (hunk 7 above)
Comment (nirinchev) [reply src/tools/mongodb/create/createIndex.ts:82]: The `typeSets` functionality is in preview.

File: src/tools/mongodb/create/createIndex.ts (hunk 5 above)
Comment (nirinchev) [reply src/tools/mongodb/create/createIndex.ts:72]: Discussing with the search team, `searchAnalyzer` and `analyzers` is not something we need to support based on current usage patterns. We can add support in the future if we see customer demand for it.

File: src/tools/mongodb/create/createIndex.ts (hunk 6 above)
Comment (nirinchev) [reply src/tools/mongodb/create/createIndex.ts:80]: * `typeSets` is still in preview, so I'm leaving it out for now.
* `numPartitions` was added
* `searchAnalyzer` and `analyzers` are out of scope per search team's recommendation

I'm leaving the following out of this PR for now as they have more complex schema and have been more error prone when testing:
* `synonyms`
* `storedSources`

We can add them as we move toward GA and build out a more comprehensive testing suite.

File: src/tools/mongodb/create/createIndex.ts (hunk 8 above)
Comment (nirinchev) [reply src/tools/mongodb/create/createIndex.ts:93]: Leaving this out of scope for now and we can decide how much detail to provide as we build out a more comprehensive testing suite.
